File: taurus-protect-sdk-python/taurus_protect/models/pagination.py
"""Pagination values and the builders every list method goes through.

Offset lists return a :class:`Pagination`, cursor lists a :class:`CursorPage`. Both are
built here and nowhere else, so a list cannot drift from the shared contract (repo-root
``CLAUDE.md`` -> "Pagination (cross-SDK)").

validatord omits zero values, so an absent ``totalItems``/``offset``/``hasNext``/``next``
means 0/false/none and ``{}`` is a valid empty page.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Dict, Optional, Union

from pydantic import BaseModel, Field

from taurus_protect.errors import APIError

DEFAULT_PAGE_SIZE = 20
MAX_PAGE_SIZE = 100

# Price history cannot page: the whole series comes back in one reply.
PRICE_HISTORY_MAX_LIMIT = 365

# The largest count every SDK represents exactly (TypeScript numbers are doubles).
MAX_COUNT = 2**53 - 1

# How each offset endpoint computes the NEXT page's offset.
REPLY_OFFSET = "reply_offset"
PLUS_ROWS = "plus_rows"
PLUS_MIN_ROWS_LIMIT = "plus_min_rows_limit"
PLUS_SERVER_ROWS = "plus_server_rows"
PLUS_LIMIT = "plus_limit"

OFFSET_RULES = (REPLY_OFFSET, PLUS_ROWS, PLUS_MIN_ROWS_LIMIT, PLUS_SERVER_ROWS, PLUS_LIMIT)

# Operations that cannot page: one reply, bounded by a limit.
LIMIT_ONLY = "limit_only"

_PAGE_REQUEST_NEXT = "NEXT"
_CANONICAL_COUNT = re.compile(r"(?:0|[1-9][0-9]*)\Z")


class Pagination(BaseModel):
    """
    The page window of an offset list. Never None on success.

    Continue with ``offset=next_offset`` while ``has_more`` is true.

    Attributes:
        limit: The page size that was sent.
        offset: The offset that was sent (never the reply's).
        total_items: The server's total, reduced by rows the SDK withheld.
        next_offset: The offset of the next page, in the server's numbering.
        has_more: Whether another page exists.
    """

    limit: int = Field(default=0, description="Page size that was sent")
    offset: int = Field(default=0, description="Offset that was sent")
    total_items: int = Field(default=0, description="Server total minus withheld rows")
    next_offset: int = Field(default=0, description="Offset of the next page")
    has_more: bool = Field(default=False, description="Whether another page exists")

    model_config = {"frozen": True}


class CursorPage(BaseModel):
    """
    The page window of a cursor list. Never None on success.

    Continue with ``cursor=next_cursor`` while ``has_more`` is true.

    Attributes:
        page_size: The page size that was sent.
        next_cursor: Opaque cursor of the next page; empty when ``has_more`` is false.
        has_more: Whether another page exists.
        total_items: The server's total, only on endpoints that report one.
    """

    page_size: int = Field(default=0, description="Page size that was sent")
    next_cursor: str = Field(default="", description="Cursor of the next page, or empty")
    has_more: bool = Field(default=False, description="Whether another page exists")
    total_items: Optional[int] = Field(
        default=None, description="Server total, only where the endpoint reports one"
    )

    model_config = {"frozen": True}


@dataclass(frozen=True)
class CursorRequest:
    """The cursor fields one page request sends. Build it with :func:`cursor_request`."""

    page_size: int
    current_page: Optional[str] = None
    page_request: Optional[str] = None

    def query_params(self, prefix: str = "cursor") -> Dict[str, Optional[str]]:
        """Keyword arguments for a generated operation taking ``<prefix>_*`` parameters."""
        return {
            f"{prefix}_current_page": self.current_page,
            f"{prefix}_page_request": self.page_request,
            f"{prefix}_page_size": str(self.page_size),
        }


class CursorListOptions(BaseModel):
    """
    The page window of a cursor list's options.

    Pass ``cursor`` (a previous ``next_cursor``) to continue a walk, or the low-level
    ``current_page``/``page_request`` pair; combining them is refused.
    """

    page_size: Optional[int] = Field(default=None, description="Page size (default 20, max 100)")
    cursor: Optional[str] = Field(default=None, description="A previous next_cursor")
    current_page: Optional[str] = Field(default=None, description="Low-level page token")
    page_request: Optional[str] = Field(
        default=None, description="Low-level page direction (FIRST, PREVIOUS, NEXT, LAST)"
    )

    def to_cursor_request(self) -> "CursorRequest":
        """Resolve these options into the cursor fields one request sends."""
        return cursor_request(
            self.page_size,
            self.cursor,
            current_page=self.current_page,
            page_request=self.page_request,
        )


def resolve_page_size(
    value: Optional[int],
    name: str = "page_size",
    maximum: Optional[int] = MAX_PAGE_SIZE,
) -> int:
    """
    Resolve a page size to the value sent: unset or 0 means the default.

    Args:
        value: The caller's page size.
        name: The option name, used in the error.
        maximum: The largest accepted value, or None for no SDK maximum.

    Returns:
        The page size to send.

    Raises:
        ValueError: If the value is negative, above ``maximum``, or not an integer.
    """
    if value is None:
        return DEFAULT_PAGE_SIZE
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{name} must be an integer, got {value!r}")
    if value == 0:
        return DEFAULT_PAGE_SIZE
    if value < 0:
        if maximum is None:
            raise ValueError(f"{name} cannot be negative, got {value}")
        raise ValueError(f"{name} must be between 1 and {maximum}, got {value}")
    if maximum is not None and value > maximum:
        raise ValueError(f"{name} must be at most {maximum}, got {value}")
    return value


def resolve_offset(value: Optional[int], name: str = "offset") -> int:
    """
    Resolve an offset: unset means 0.

    Raises:
        ValueError: If the value is negative or not an integer.
    """
    if value is None:
        return 0
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{name} must be an integer, got {value!r}")
    if value < 0:
        raise ValueError(f"{name} cannot be negative, got {value}")
    return value


def offset_query(limit: int, offset: int) -> Dict[str, Optional[str]]:
    """The ``limit``/``offset`` query arguments; offset 0 is the server default and is omitted."""
    return {"limit": str(limit), "offset": str(offset) if offset else None}


def parse_count(raw: Optional[Union[str, int]], field: str) -> int:
    """
    Parse a count the server sent; absent means 0.

    A count that is not a canonical decimal in ``[0, 2**53 - 1]`` is refused rather than
    read as 0: a silent 0 would end every walk early.

    Raises:
        APIError: If the count is malformed or out of range.
    """
    if raw is None:
        return 0
    if isinstance(raw, bool):
        raise _invalid_reply(f"{field} is not a count: {raw!r}")
    if isinstance(raw, int):
        value = raw
    elif isinstance(raw, str) and _CANONICAL_COUNT.match(raw):
        value = int(raw)
    else:
        raise _invalid_reply(f"{field} is not a canonical count: {raw!r}")
    if value < 0 or value > MAX_COUNT:
        raise _invalid_reply(f"{field} is out of range: {raw!r}")
    return value


def offset_pagination(
    rule: str,
    *,
    limit: int,
    offset: int,
    served_rows: int,
    total_items: Optional[Union[str, int]],
    reply_offset: Optional[Union[str, int]] = None,
    excluded: int = 0,
) -> Pagination:
    """
    Build the :class:`Pagination` of one offset-list reply.

    Args:
        rule: How the endpoint computes the next offset (one of ``OFFSET_RULES``).
        limit: The page size that was sent.
        offset: The offset that was sent.
        served_rows: Rows the SERVER returned, before any SDK exclusion.
        total_items: The reply's ``totalItems``, as received.
        reply_offset: The reply's ``offset`` (``reply_offset`` rule only), as received.
        excluded: Rows the SDK withheld from the caller.

    Returns:
        The page window. ``next_offset`` and ``has_more`` use the server's numbers;
        only ``total_items`` is reduced by ``excluded``.

    Raises:
        APIError: If a count in the reply is malformed.
        ValueError: If ``rule`` is unknown.
    """
    total = parse_count(total_items, "totalItems")
    if rule == REPLY_OFFSET:
        # validatord's reply offset is already offset + rows: it is the NEXT page.
        if reply_offset is None:
            next_offset = offset + served_rows
        else:
            next_offset = parse_count(reply_offset, "offset")
    elif rule in (PLUS_ROWS, PLUS_SERVER_ROWS):
        next_offset = offset + served_rows
    elif rule == PLUS_MIN_ROWS_LIMIT:
        # A synthetic daemon user / tech group can be appended beyond the limit.
        next_offset = offset + min(served_rows, limit)
    elif rule == PLUS_LIMIT:
        # Skipped rows keep their SQL slot, so a short page is not the end.
        next_offset = offset + limit
    else:
        raise ValueError(f"unknown pagination rule {rule!r}")

    return Pagination(
        limit=limit,
        offset=offset,
        total_items=max(0, total - excluded),
        # No progress ends a walk; some totals are upper bounds.
        next_offset=next_offset,
        has_more=offset < next_offset < total,
    )


def cursor_request(
    page_size: Optional[int] = None,
    cursor: Optional[str] = None,
    *,
    current_page: Optional[str] = None,
    page_request: Optional[str] = None,
    name: str = "page_size",
) -> CursorRequest:
    """
    Resolve the cursor fields of one cursor-list request.

    ``cursor`` is a previous :attr:`CursorPage.next_cursor` and sends
    ``currentPage=<cursor>`` + ``pageRequest=NEXT``. ``current_page``/``page_request``
    are the low-level alternative; combining them with ``cursor`` is refused.

    Raises:
        ValueError: If the page size is invalid, or ``cursor`` is combined with the
            low-level options.
    """
    size = resolve_page_size(page_size, name)
    if cursor is not None and not isinstance(cursor, str):
        raise ValueError(f"cursor must be a string, got {type(cursor).__name__}")
    if cursor:
        if current_page or page_request:
            raise ValueError(
                "cursor cannot be combined with current_page or page_request: "
                "pass a previous next_cursor, or the low-level pair, not both"
            )
        return CursorRequest(page_size=size, current_page=cursor, page_request=_PAGE_REQUEST_NEXT)
    return CursorRequest(
        page_size=size,
        current_page=current_page or None,
        page_request=page_request or None,
    )


def cursor_page(
    page_size: int,
    reply_cursor: Any = None,
    *,
    token: Optional[Union[str, bytes]] = None,
    total: Optional[Union[str, int]] = None,
    has_total: bool = False,
    excluded: int = 0,
) -> CursorPage:
    """
    Build the :class:`CursorPage` of one cursor-list reply.

    Cursor lists pass the reply's generated ``cursor`` (``currentPage``/``hasNext``);
    the two token-only operations (wallet tokens ``next``, rules history ``cursor``)
    pass ``token`` instead.

    Args:
        page_size: The page size that was sent.
        reply_cursor: The reply's ``cursor`` model, or None when the reply omitted it.
        token: The reply's continuation token (token-only operations).
        total: The reply's total, as received (only where ``has_total``).
        has_total: Whether this endpoint reports a total.
        excluded: Rows the SDK withheld, subtracted from the total.

    Returns:
        The page window; ``next_cursor`` is the reply's ``currentPage`` only when
        ``hasNext``. A cursor derived any other way re-reads page 1 or is a 400.

    Raises:
        APIError: If the reply's total is malformed, or ``hasNext`` has no cursor.
    """
    total_items: Optional[int] = None
    if has_total:
        total_items = max(0, parse_count(total, "total") - excluded)

    if token is not None and reply_cursor is not None:
        raise ValueError("pass the reply cursor or the reply token, not both")

    if token is not None:
        text = token.decode("utf-8") if isinstance(token, bytes) else token
        return CursorPage(
            page_size=page_size,
            next_cursor=text,
            has_more=bool(text),
            total_items=total_items,
        )

    has_more = bool(reply_cursor is not None and reply_cursor.has_next)
    next_cursor = ""
    if has_more:
        next_cursor = reply_cursor.current_page or ""
        if not next_cursor:
            raise _invalid_reply("cursor.hasNext is true but cursor.currentPage is empty")
    return CursorPage(
        page_size=page_size,
        next_cursor=next_cursor,
        has_more=has_more,
        total_items=total_items,
    )


def _invalid_reply(detail: str) -> APIError:
    return APIError(f"invalid list reply: {detail}", description="Invalid Reply")

File: taurus-protect-sdk-typescript/src/models/pagination.ts
/**
 * Pagination for list results.
 *
 * Implements the cross-SDK contract in the repo-root CLAUDE.md § "Pagination (cross-SDK)",
 * pinned in every SDK by `scripts/resources/pagination-vectors.json`. Every list method
 * builds its paging request and its paging result through the functions in this file, so
 * page sizes are validated, cursors forwarded and next offsets computed in one place.
 *
 * validatord omits zero values, so an absent `totalItems` / `offset` / `hasNext` / `next`
 * means 0 / false / none, and `{}` is a valid empty page.
 */

import { PaginationError, ValidationError } from "../errors";

/** Page size sent when the caller sets none, or 0. */
export const DEFAULT_PAGE_SIZE = 20;

/** Largest page size a list method accepts. */
export const MAX_PAGE_SIZE = 100;

/** Largest `limit` the price history accepts: a year of daily points. */
export const MAX_PRICE_HISTORY_LIMIT = 365;

/** Largest count a reply may carry: 2^53 - 1, the largest integer a number holds exactly. */
export const MAX_COUNT = Number.MAX_SAFE_INTEGER;

/**
 * Offset pagination of a list result. Present on every successful offset list, the empty
 * page included.
 */
export interface Pagination {
  /** The page size that was sent. */
  readonly limit: number;
  /** The offset that was sent — the request's, never the reply's. */
  readonly offset: number;
  /** The server's total, reduced by rows this SDK withheld from the page (never below 0). */
  readonly totalItems: number;
  /** Offset of the next page: pass it back as `offset` while `hasMore` is true. */
  readonly nextOffset: number;
  /** Whether another page exists. */
  readonly hasMore: boolean;
}

/**
 * Cursor pagination of a list result. Present on every successful cursor list, the empty
 * page included.
 */
export interface CursorPage {
  /** The page size that was sent. */
  readonly pageSize: number;
  /** Cursor of the next page: pass it back as `cursor` while `hasMore` is true; "" otherwise. */
  readonly nextCursor: string;
  /** Whether another page exists. */
  readonly hasMore: boolean;
  /** The server's total where the endpoint returns one (0 when omitted); undefined otherwise. */
  readonly totalItems?: number;
}

/** A page of an offset list. */
export interface PaginatedResult<T> {
  /** The rows of this page. */
  readonly items: T[];
  /** Where this page sits and how to fetch the next one. */
  readonly pagination: Pagination;
}

/** Direction of the low-level cursor navigation options. */
export type PageRequest = "FIRST" | "PREVIOUS" | "NEXT" | "LAST";

/** Paging options of an offset list. */
export interface OffsetPageOptions {
  /** Page size, 1 to 100 (default 20). */
  readonly limit?: number;
  /** Rows to skip (default 0): a previous result's `pagination.nextOffset`. */
  readonly offset?: number;
}

/** Paging options of a cursor list. */
export interface CursorPageOptions {
  /** Page size, 1 to 100 (default 20). */
  readonly pageSize?: number;
  /** A previous result's `pagination.nextCursor`: fetches the page after it. */
  readonly cursor?: string;
}

/** Paging options of a cursor list that also offers low-level cursor navigation. */
export interface CursorNavigationOptions extends CursorPageOptions {
  /** Low-level: the page token to navigate from. Cannot be combined with `cursor`. */
  readonly currentPage?: string;
  /** Low-level: the direction to navigate from `currentPage`. Cannot be combined with `cursor`. */
  readonly pageRequest?: PageRequest;
}

/**
 * Which bound a page size is checked against: a pageable list (max 100), the price history
 * (max 365), or an export that cannot page (no maximum — a larger limit is the only way to
 * reach more rows).
 */
export type PageSizeKind = "page" | "price_history" | "export";

/**
 * How an offset endpoint's next page is found (the contract's rule table).
 *
 * - `reply_offset`: the reply's `offset`, else offset + rows (wallets, addresses).
 * - `plus_rows`: offset + rows (transactions, fee payers, actions).
 * - `plus_min_rows_limit`: offset + min(rows, limit) — the server may append a synthetic
 *   daemon user / technical group beyond `limit` (users, groups).
 * - `plus_server_rows`: offset + the rows the SERVER returned, before this SDK withheld
 *   any (whitelisted addresses).
 * - `plus_limit`: offset + limit — skipped rows keep their SQL slot (whitelisted contracts).
 */
export type OffsetRule =
  | "reply_offset"
  | "plus_rows"
  | "plus_min_rows_limit"
  | "plus_server_rows"
  | "plus_limit";

/** The paging request actually sent to an offset endpoint. */
export interface OffsetRequest {
  readonly limit: number;
  readonly offset: number;
}

/** An offset reply's paging fields, exactly as the generated reply types carry them. */
export interface OffsetReply {
  readonly totalItems?: string | number;
  readonly offset?: string | number;
}

/** The paging request actually sent to a cursor endpoint. */
export interface CursorRequest {
  /** The page size that is sent. */
  readonly pageSize: number;
  /**
   * The cursor in the shape of the generated `TgvalidatordRequestCursor`, which is the
   * body form; the `cursor.*` and `requestCursor.*` query forms are spelled from it.
   */
  readonly cursor: {
    readonly currentPage?: string;
    readonly pageRequest?: string;
    readonly pageSize: string;
  };
}

/** The paging request actually sent to a token endpoint: `limit` plus the opaque token. */
export interface TokenRequest {
  /** The page size that is sent. */
  readonly pageSize: number;
  /** The page size as the `limit` query string. */
  readonly limit: string;
  /** The token to continue from, when the caller passed one. */
  readonly cursor?: string;
}

/** A reply cursor's fields, exactly as the generated `TgvalidatordResponseCursor` carries them. */
export interface ReplyCursor {
  readonly currentPage?: string;
  readonly hasNext?: boolean;
}

/** The total of a cursor or token page, for the endpoints that return one. */
export interface PageTotal {
  /** The reply's total as the generated reply type carries it; absent means 0. */
  readonly total: string | number | undefined;
  /** Rows this SDK withheld from the page; they reduce `totalItems` (never below 0). */
  readonly excluded?: number;
}

/**
 * Resolves a requested page size: unset or 0 means the default, anything else must be a
 * whole number in bounds. The one page-size validator every list method goes through.
 *
 * @param value - The page size the caller passed
 * @param option - The option's name, for the error message
 * @param kind - `page` (max 100), `price_history` (max 365) or `export` (no maximum)
 * @returns The page size to send
 * @throws {@link ValidationError} If the value is negative, fractional or above the maximum
 */
export function resolvePageSize(
  value: number | null | undefined,
  option = "pageSize",
  kind: PageSizeKind = "page"
): number {
  if (value === undefined || value === null || value === 0) {
    return DEFAULT_PAGE_SIZE;
  }
  if (typeof value !== "number" || !Number.isSafeInteger(value)) {
    throw new ValidationError(`${option} must be a whole number, got ${String(value)}`);
  }
  if (value < 0) {
    throw new ValidationError(`${option} must not be negative, got ${value}`);
  }
  const max =
    kind === "page" ? MAX_PAGE_SIZE : kind === "price_history" ? MAX_PRICE_HISTORY_LIMIT : undefined;
  if (max !== undefined && value > max) {
    throw new ValidationError(`${option} must be at most ${max}, got ${value}`);
  }
  return value;
}

/**
 * Resolves a requested offset: unset means 0, anything else must be a whole number >= 0.
 *
 * @param value - The offset the caller passed
 * @param option - The option's name, for the error message
 * @returns The offset to send
 * @throws {@link ValidationError} If the value is negative or fractional
 */
export function resolveOffset(value: number | null | undefined, option = "offset"): number {
  if (value === undefined || value === null) {
    return 0;
  }
  if (typeof value !== "number" || !Number.isSafeInteger(value)) {
    throw new ValidationError(`${option} must be a whole number, got ${String(value)}`);
  }
  if (value < 0) {
    throw new ValidationError(`${option} must not be negative, got ${value}`);
  }
  return value;
}

/**
 * Resolves the paging request of an offset list.
 *
 * @param options - The caller's `limit` / `offset`
 * @returns The limit and offset to send
 * @throws {@link ValidationError} If either is out of bounds
 */
export function offsetRequest(options: OffsetPageOptions | undefined): OffsetRequest {
  return {
    limit: resolvePageSize(options?.limit, "limit"),
    offset: resolveOffset(options?.offset, "offset"),
  };
}

/**
 * Resolves the paging request of a cursor list.
 *
 * With `cursor` set the request continues after that page: `currentPage=<cursor>` and
 * `pageRequest=NEXT`. Without it the low-level `currentPage` / `pageRequest` pass through
 * unchanged. The page size is always sent.
 *
 * @param options - The caller's paging options
 * @returns The page size and the cursor to send
 * @throws {@link ValidationError} If the page size is out of bounds, or `cursor` is combined
 *   with `currentPage` or `pageRequest`
 */
export function cursorRequest(options: CursorNavigationOptions | undefined): CursorRequest {
  const pageSize = resolvePageSize(options?.pageSize, "pageSize");
  const cursor = continuationToken(options?.cursor);
  if (cursor !== undefined) {
    if (options?.currentPage !== undefined) {
      throw new ValidationError(
        "cursor and currentPage cannot both be set: cursor already names the page to continue from"
      );
    }
    if (options?.pageRequest !== undefined) {
      throw new ValidationError(
        "cursor and pageRequest cannot both be set: a cursor always continues with the NEXT page"
      );
    }
    return {
      pageSize,
      cursor: { currentPage: cursor, pageRequest: "NEXT", pageSize: String(pageSize) },
    };
  }
  return {
    pageSize,
    cursor: {
      currentPage: options?.currentPage,
      pageRequest: options?.pageRequest,
      pageSize: String(pageSize),
    },
  };
}

/**
 * Resolves the paging request of a token list (`limit` plus an opaque `cursor` token).
 *
 * @param options - The caller's page size and token
 * @returns The page size, limit and token to send
 * @throws {@link ValidationError} If the page size is out of bounds
 */
export function tokenRequest(options: CursorPageOptions | undefined): TokenRequest {
  const pageSize = resolvePageSize(options?.pageSize, "pageSize");
  return { pageSize, limit: String(pageSize), cursor: continuationToken(options?.cursor) };
}

/**
 * Builds the pagination of an offset list page — the one place a next offset is computed.
 *
 * `offset` is the request's. `nextOffset` follows the endpoint's rule, and `hasMore` is
 * `nextOffset > offset && nextOffset < total`: no progress ends a walk, and some totals are
 * upper bounds. `totalItems` is the server total less the rows withheld, never below 0;
 * `nextOffset` and `hasMore` always use the server's numbers.
 *
 * @param rule - The endpoint's next-offset rule
 * @param request - The limit and offset that were sent
 * @param reply - The reply's `totalItems` / `offset`, as the generated reply carries them
 * @param servedRows - How many rows the server returned, before this SDK withheld any
 * @param sdkExcluded - How many of those rows this SDK withheld
 * @returns The page's pagination
 * @throws {@link PaginationError} If the reply's total or offset is not a canonical count
 */
export function buildOffsetPagination(
  rule: OffsetRule,
  request: OffsetRequest,
  reply: OffsetReply,
  servedRows: number,
  sdkExcluded = 0
): Pagination {
  const total = parseReplyCount(reply.totalItems, "totalItems");
  const { limit, offset } = request;
  let nextOffset: number;
  switch (rule) {
    case "reply_offset":
      nextOffset =
        reply.offset === undefined || reply.offset === null
          ? offset + servedRows
          : parseReplyCount(reply.offset, "offset");
      break;
    case "plus_rows":
    case "plus_server_rows":
      nextOffset = offset + servedRows;
      break;
    case "plus_min_rows_limit":
      nextOffset = offset + Math.min(servedRows, limit);
      break;
    case "plus_limit":
      nextOffset = offset + limit;
      break;
    default:
      throw new ValidationError(`unknown offset rule: ${describe(rule)}`);
  }
  return {
    limit,
    offset,
    totalItems: Math.max(0, total - sdkExcluded),
    nextOffset,
    hasMore: nextOffset > offset && nextOffset < total,
  };
}

/**
 * Builds the pagination of a cursor or token list page — the one place a next cursor is
 * derived.
 *
 * `next` is the reply's cursor object for a cursor list (`hasMore` = `hasNext`, and the
 * next cursor is its `currentPage`), or the reply's token for a token list (`hasMore` =
 * the token is present, and the next cursor is the token). Either way the next cursor is
 * "" when there is no next page, and is never derived any other way: sending NEXT past the
 * end is a 400 on the v2 lists and silently returns page 1 on the keyset lists.
 *
 * @param pageSize - The page size that was sent
 * @param next - The reply's cursor object, or its token
 * @param total - The reply's total, for the endpoints that return one
 * @returns The page's pagination
 * @throws {@link PaginationError} If the cursor reports a next page without naming it, or
 *   the total is not a canonical count
 */
export function buildCursorPage(
  pageSize: number,
  next: ReplyCursor | string | undefined | null,
  total?: PageTotal
): CursorPage {
  let nextCursor = "";
  let hasMore = false;
  if (typeof next === "string") {
    hasMore = next.length > 0;
    nextCursor = next;
  } else if (next !== undefined && next !== null) {
    hasMore = next.hasNext === true;
    if (hasMore) {
      if (typeof next.currentPage !== "string" || next.currentPage.length === 0) {
        throw new PaginationError(
          "reply cursor reports a next page (hasNext) but carries no currentPage to continue from"
        );
      }
      nextCursor = next.currentPage;
    }
  }
  if (total === undefined) {
    return { pageSize, nextCursor, hasMore };
  }
  const count = parseReplyCount(total.total, "total");
  return {
    pageSize,
    nextCursor,
    hasMore,
    totalItems: Math.max(0, count - (total.excluded ?? 0)),
  };
}

/**
 * Parses a count off a reply: a canonical decimal in [0, 2^53 - 1]. Absent means 0.
 *
 * Anything else raises rather than reading as 0, which would end a walk early and pass a
 * truncated list off as complete.
 *
 * @param value - The count as the generated reply type carries it
 * @param field - The reply field's name, for the error message
 * @returns The count
 * @throws {@link PaginationError} If the value is not a canonical count
 */
export function parseReplyCount(value: string | number | undefined | null, field: string): number {
  if (value === undefined || value === null) {
    return 0;
  }
  if (typeof value === "number") {
    if (Number.isSafeInteger(value) && value >= 0) {
      return value;
    }
    throw new PaginationError(`reply ${field} is not a count in [0, 2^53 - 1]: ${value}`);
  }
  if (typeof value !== "string" || !/^(0|[1-9][0-9]*)$/.test(value)) {
    throw new PaginationError(`reply ${field} is not a canonical count: ${describe(value)}`);
  }
  const count = Number(value);
  if (!Number.isSafeInteger(count)) {
    throw new PaginationError(`reply ${field} ${describe(value)} is above 2^53 - 1`);
  }
  return count;
}

/** Normalises a caller's continuation token: "" and unset both mean "start at page 1". */
function continuationToken(cursor: string | undefined): string | undefined {
  if (cursor === undefined || cursor === null) {
    return undefined;
  }
  if (typeof cursor !== "string") {
    throw new ValidationError("cursor must be a string: a previous result's nextCursor");
  }
  return cursor.length === 0 ? undefined : cursor;
}

/** Quotes a server value for an error message, bounded so a large reply cannot flood it. */
function describe(value: unknown): string {
  const text = JSON.stringify(value) ?? String(value);
  return text.length > 40 ? `${text.slice(0, 37)}...` : text;
}
